js/source/geojson_source.js

```diff
@@ -95,82 +90,52 @@ GeoJSONSource.prototype = util.inherit(Evented, /** @lends GeoJSONSource.prototy
         this._data = data;
         this._dirty = true;
 
-        this.fire('change');
-
-        if (this.map)
-            this.update(this.map.transform);
+        this._onChange();
 
         return this;
     },
 
-    onAdd: function(map) {
-        this.map = map;
-    },
-
-    loaded: function() {
-        return this._loaded && this._pyramid.loaded();
-    },
-
-    update: function(transform) {
-        if (this._dirty) {
-            this._updateData();
-        }
-
-        if (this._loaded) {
-            this._pyramid.update(this.used, transform);
-        }
-    },
-
-    reload: function() {
-        if (this._loaded) {
-            this._pyramid.reload();
-        }
-    },
-
-    serialize: function() {
-        return {
-            type: 'geojson',
-            data: this._data
-        };
-    },
-
-    getVisibleCoordinates: Source._getVisibleCoordinates,
-    getTile: Source._getTile,
-
-    queryRenderedFeatures: Source._queryRenderedVectorFeatures,
-    querySourceFeatures: Source._querySourceFeatures,
-
-    _updateData: function() {
+    _updateData: function(cb) {
         this._dirty = false;
-        var options = {
-            tileSize: this.tileSize,
+        var options = util.extend({
             source: this.id,
-            geojsonVtOptions: this.geojsonVtOptions,
-            cluster: this.cluster,
-            superclusterOptions: this.superclusterOptions
-        };
+            extent: EXTENT,
+            scale: EXTENT / this.tileSize,
+            minZoom: this.minzoom,
+            maxZoom: this.maxzoom
+        }, this.options);
 
         var data = this._data;
         if (typeof data === 'string') {
             options.url = typeof window != 'undefined' ? urlResolve(window.location.href, data) : data;
         } else {
             options.data = JSON.stringify(data);
         }
-        this.workerID = this.dispatcher.send('parse geojson', options, function(err) {
+        this.workerID = this.dispatcher.send(this.type + '.parse', options, function(err) {
+            // TODO: not quite sure if we need this anymore
             this._loaded = true;
+            cb(err);
+
+        }.bind(this));
+    },
+
+    load: function (tile, callback) {
+        if (!this._dirty) {
+            return this._load(tile, callback);
+        }
+
+        this._updateData(function (err) {
             if (err) {
-                this.fire('error', {error: err});
-            } else {
-                this._pyramid.reload();
-                this.fire('change');
+                return callback(err);
             }
-
+            this._load(tile, callback);
         }.bind(this));
     },
 
-    _loadTile: function(tile) {
+    _load: function(tile, callback) {
```

Comment:
`GeoJSONSource#_load` and `GeoJSONSource#load` could be named more clearly for what they do. It seems like `loadTile` and `loadTileUncached` would be appropriate names. The important part is that we differentiate the notion of loading tiles from the notion of the source being loaded.